### ISSUE TITLE
Storybook Storyshots errors with "Your test suite must contain at least one test."

### ISSUE THREAD

**Issue by DaveMBush** [nrwl/nx#8242]
## Current Behavior
When I run tests with storyshots the test(s) fail with "Your test suite must contain at least one test"

## Expected Behavior
I expect it to run my tests

This has worked for me using Angular 12 and Webpack4 and storybook 6.2.x but no longer works with storybook 6.3.x or greater (including the current 6.5.alpha4 version) and webpack5.  (I'm not sure that the webpack4/5 is the issue but including for completeness.

## Steps to Reproduce

Steps to reproduce storyshot error using Angular ~13.0.0 and storybook ~6.4.5

- Create new NX Workspace with Angular

``` shell
npx create-nx-workspace --preset=angular
```

Respond to the questions:

- workspace=demo-workspace
- application=demo
- CSS
- no Cloud

- Add Storybook plugin to NX

``` shell
npm i -D @nrwl/storybook
```

- Add Storybook nrwl schematic

``` shell
npm i -D @nrwl/storybook
```

- Add Storybook to demo app

``` shell
npx nx g @nrwl/angular:storybook-configuration demo
```

Answer the questions:
- cypress=n
- Auto generate stories=y
- Auto generate specs=n

- There is a known issue that requires us to rollback the angular version from ~13.1.0 to ~13.0.0.  Do so and run `npm install` again.

- At this point you should be able to run storybook

``` shell
npx nx run demo:storybook
```

- Add storyshots to the project

``` shell
npm install @storybook/addon-storyshots
```

- Add storyshots.spec.js to apps/demo/src and put the standard two lines in the file:

``` javascript
import initStoryShots from '@storybook/addon-storyshots';

initStoryShots();
```

- Add the following sections to jest.config.js file in /apps/demo

``` javascript
  moduleFileExtensions: ['ts', 'js', 'html', 'json', 'mjs'],
  moduleNameMapper: {
    'jest-preset-angular/build/setup-jest': 'jest-preset-angular/setup-jest',
    'jest-preset-angular/build/AngularNoNgAttributesSnapshotSerializer':
      'jest-preset-angular/build/serializers/no-ng-attributes',
    'jest-preset-angular/build/AngularSnapshotSerializer':
      'jest-preset-angular/build/serializers/ng-snapshot',
    'jest-preset-angular/build/HTMLCommentSerializer':
      'jest-preset-angular/build/serializers/html-comment'
  },
```

This solves the spdx* issue that you would otherwise see and maps functions that no longer exists to where they've been moved to.

- At this point, you should be able to run the tests and see the error

``` shell
npm run test
```

Errors on storyshots.spec.js with:
"Your test suite must contain at least one test."

Storybook is not part of the examples

https://github.com/DaveMBush/demo-workspace

### Environment

Node : 16.13.1
  OS   : win32 x64
  npm  : 8.3.0

  nx : 13.3.9
  @nrwl/angular : 13.3.9
  @nrwl/cli : 13.3.9
  @nrwl/cypress : 13.3.9
  @nrwl/devkit : 13.3.9
  @nrwl/eslint-plugin-nx : 13.3.9
  @nrwl/express : undefined
  @nrwl/jest : 13.3.9
  @nrwl/linter : 13.3.9
  @nrwl/nest : undefined
  @nrwl/next : undefined
  @nrwl/node : undefined
  @nrwl/nx-cloud : undefined
  @nrwl/react : undefined
  @nrwl/react-native : undefined
  @nrwl/schematics : undefined
  @nrwl/tao : 13.3.9
  @nrwl/web : undefined
  @nrwl/workspace : 13.3.9
  @nrwl/storybook : 13.3.9
  @nrwl/gatsby : undefined
  typescript : 4.4.4
  rxjs : 7.4.0
  ---------------------------------------
  Community plugins:
         @angular/animations: 13.0.3
         @angular/common: 13.0.3
         @angular/compiler: 13.0.3
         @angular/core: 13.0.3
         @angular/forms: 13.0.3
         @angular/platform-browser: 13.0.3
         @angular/platform-browser-dynamic: 13.0.3
         @angular/router: 13.0.3
         @angular-devkit/build-angular: 13.0.4
         @angular/cli: 13.0.4
         @angular/compiler-cli: 13.0.3
         @angular/language-service: 13.0.3
         @storybook/angular: 6.4.9

(node:13888) [DEP0148] DeprecationWarning: Use of deprecated folder mapping "./" in the "exports" field module resolution of the package at C:\Users\davem\documents\code\demo-workspace\node_modules\tslib\package.json.
Update this package.json to use a subpath pattern like "./*".
(Use `node --trace-deprecation ...` to show where the warning was created)

**Comment by DaveMBush** [comment]
With the help of the Storybook team I have more information on this...

if I cd into the app project

``` shell
cd apps/demo
```

And then run jest directly

``` shell
npx jest
```

The stories run as expected.  So something about running via nx is causing this to not work?

**Comment by asinitson** [comment]
Confirming that the same issue is repeatable with `nx` 12.* and Angular 12.*. All the steps described above also apply to these versions (a lot of workarounds and manual configuration tweaking!).

From nx workspace root directory:
```
$ npx jest
 PASS   hello-nx  apps/hello-nx/src/app/foo/foo.component.spec.ts
 PASS   hello-nx  apps/hello-nx/src/app/app.component.spec.ts
 FAIL   hello-nx  apps/hello-nx/snapshots/snapshots.spec.ts
  ● Test suite failed to run

    storyshots found 0 stories

      1 | import initStoryshots from '@storybook/addon-storyshots';
      2 |
    > 3 | initStoryshots();
        |               ^
      4 |

      at Object.testStorySnapshots [as default] (../../node_modules/@storybook/addon-storyshots/dist/ts3.9/api/index.js:86:15)
      at Object.<anonymous> (snapshots/snapshots.spec.ts:3:15)

Test Suites: 1 failed, 2 passed, 3 total
Tests:       4 passed, 4 total
Snapshots:   0 total
Time:        3.364 s
Ran all test suites.
```

```
$ APP_NAME=hello-nx
cd apps/${APP_NAME}
$ npx jest
 PASS   hello-nx  src/app/foo/foo.component.spec.ts
 PASS   hello-nx  src/app/app.component.spec.ts
(node:9141) DeprecationWarning: `storyFn` is deprecated and will be removed in Storybook 7.0.

https://github.com/storybookjs/storybook/blob/next/MIGRATION.md#deprecated-storyfn
(Use `node --trace-deprecation ...` to show where the warning was created)
(node:9141) DeprecationWarning: `component` story return value is deprecated, and will be removed in Storybook 7.0.
Instead, use `export const default = () => ({ component: AppComponent });`
or
```
export const Primary: Story = () => ({});
Primary.parameters = { component: AppComponent };
```
Read more at 
- https://github.com/storybookjs/storybook/blob/next/MIGRATION.md#deprecated-angular-story-component).
- https://storybook.js.org/docs/angular/writing-stories/parameters
 PASS   hello-nx  snapshots/snapshots.spec.ts
 › 2 snapshots written.

Snapshot Summary
 › 2 snapshots written from 1 test suite.

Test Suites: 3 passed, 3 total
Tests:       6 passed, 6 total
Snapshots:   2 written, 2 total
Time:        3.489 s
Ran all test suites.
```

**Comment by DaveMBush** [comment]
@asinitson are you using Jest 27?

Could you also post in the storybook github https://github.com/storybookjs/storybook/issues/17047 as I believe, at this point, the issue is between Jest 27 and storyshots (unless you are using Jest 26 and then that theory doesn't work any more)

**Comment by mandarini** [comment]
Hi @DaveMBush ! Does your issue still exist? And is it related to Nx? If not, do you think we could close this issue, or at least mark as "blocked by third-party"?

**Comment by DaveMBush** [comment]
At this point, I'm fairly sure the problem is some interaction between Storybook and Jest 27 so we can mark it as blocked by 3rd party.

**Comment by asinitson** [comment]
> @asinitson are you using Jest 27?

Yes, it's version `27.*`. Relevant `package.json` snippet is here (as you can notice, it drags older - 6.3.* Storybook in):
```json
    "@nrwl/storybook": "12.10.1",
    "@nrwl/tao": "12.10.1",
    "@nrwl/workspace": "12.10.1",
    "@storybook/addon-essentials": "~6.3.0",
    "@storybook/addon-storyshots": "~6.3.0",
    "@storybook/angular": "~6.3.0",
    "@storybook/builder-webpack5": "~6.3.0",
    "@storybook/manager-webpack5": "~6.3.0",
    "@types/jest": "27.0.2",
    "@types/node": "14.14.33",
    "@typescript-eslint/eslint-plugin": "~4.31.1",
    "@typescript-eslint/parser": "~4.31.1",
    "cypress": "^8.3.0",
    "eslint": "7.22.0",
    "eslint-config-prettier": "8.1.0",
    "eslint-plugin-cypress": "^2.10.3",
    "jest": "27.2.3",
    "jest-preset-angular": "10.0.1",
```

I use a dummy project (I can fully publish it, if needed), I have set it up specifically to check how well does nx integrate with Storybook. I use setup close to what you described with this difference: I specify nx version during initialization (as you see yourself latest versions usually have more problems - as is the case with Angular 13.1.*, so I decided to try `12.*`):

```bash
NX_VERSION=12.10.1
APP_NAME=hello-nx
WORKSPACE=${APP_NAME}-angular
npx create-nx-workspace@${NX_VERSION} \
    ${WORKSPACE} \
    --preset=angular \
    --appName=${APP_NAME} \
    --style=css \
    --nx-cloud=false

npx nx generate storybook-configuration ${APP_NAME} \
  --configureCypress=false \
  --generateCypressSpecs=false \
  --generateStories=true
```
_The rest of the setup is the same as in the [first post](https://github.com/nrwl/nx/issues/8242#issue-1084974784)._

> Could you also post in the storybook github [storybookjs/storybook#17047](https://github.com/storybookjs/storybook/issues/17047) as I believe, at this point, the issue is between Jest 27 and storyshots (unless you are using Jest 26 and then that theory doesn't work any more)

I don't have much more to report (you already provided all the information), but I am subscribed to that issue and will post as soon as I have something. I also tried debugging a bit, but couldn't figure out at what stage do stories "get lost".

**Comment by mandarini** [comment]
Hi there! I will go ahead and close this issue, since it's been a few months, and it's a third-party dependency. If you believe that it should be kept open, let me know and I can reopen!

Thanks everyone for jumping in to help and provide workarounds! :)

**Comment by sellmeadog** [comment]
I was able to fix all of my Nx + Storybook + Storyshots issues with this configuration:

```ts
import { render } from '@testing-library/react';
import { resolve } from 'path';
import initStoryshots from '@storybook/addon-storyshots';

initStoryshots({
  configPath: resolve(__dirname, '../.storybook'),
  framework: 'react',
  renderer: render,
});
```
The crux of the issue seems to be that by default, storybook/storyshots cannot find the `.storybook` directory specific to the app/lib being tested which must be provided with the `configPath` option. Once that was set, all of the stories were discovered and I subsequently had to  configure `framework` and `renderer` to resolve some runtime issues and then everything started working as expected.

**Comment by mandarini** [comment]
Ohh nice @sellmeadog ! Thanks for the response/solution!
